File: src/data/Binary.java
package data;

public class Binary {
	/*
	 * Returns a string with the supplied targetLength as it's length value with
	 * paddings of zeros towards the left or right Supply direction = true for
	 * padding towards the left
	 */
	public static String padWithZeros(String binary, int targetLength, boolean direction) {
		if (direction)
			return padWithZerosLeft(binary, targetLength);
		return padWithZerosRight(binary, targetLength);
	}

	/*
	 * Helper function for when the value of the direction parameter is set to
	 * true.
	 */
	public static String padWithZerosLeft(String binary, int targetLength) {
		while (binary.length() < targetLength) {
			binary = "0" + binary;
		}
		return binary;
	}
	/*
	 * Helper function for when the value of the direction parameter is set to
	 * false.
	 */

	public static String padWithZerosRight(String binary, int targetLength) {
		while (binary.length() < targetLength) {
			binary += "0";
		}
		return binary;
	}

	/*
	 * Returns the supplied string of a binary represented value after
	 * performing a circular shift lift.
	 */
	public static String circularShiftLeft(String s) {
		return s.substring(1, s.length()) + s.charAt(0);
	}

	/*
	 * Returns the supplied string of a binary represented value after
	 * performing a circular shift right.
	 */
	public static String circularShiftRight(String s) {
		return s.charAt(s.length() - 1) + s.substring(0, s.length() - 1);
	}

	/*
	 * Returns the obtained value after XOR-ing the two supplied binary
	 * representations.
	 */
	public static String xor(String x, String y) {
		String z = "";
		for (int i = 0; i < x.length(); i++) {
			char c1 = x.charAt(i);
			char c2 = y.charAt(i);
			z += (c1 == c2) ? "0" : "1";
		}
		return z;
	}

	/*
	 * Converts a binary representation to it's corresponding string of bytes.
	 * Assumes 32-bit blocks with 8-bit words.
	 */
	public static String toBytes(String s) {
		String ret = "";
		for (int i = 0; i < 4; i++) {
			int val = Integer.parseInt(s.substring(8 * i, (i + 1) * 8), 2);
			ret += (char) val;
		}
		return ret;
	}

	/*
	 * Converts a string of bytes to it's corresponding binary representation.
	 */
	public static String toBinary(String s) {
		byte[] bytes = s.getBytes();
		String out = "";
		for (byte b : bytes) {
			String bString = Integer.toBinaryString(b);
			while (bString.length() < 8)
				bString = "0" + bString;
			out += bString;
		}
		return out;
	}

	public static String intToBinary(int n) {
		return Integer.toString(n, 2);
	}

	/*
	 * Converts a hexadecimal string to it's binary string representation.
	 */
	public static String binaryToHex(String hex) {
		return Integer.toString(Integer.parseInt(hex, 2), 16);
	}
}

File: python/data/Binary.py
def pad_zeros_left(bit_str, target_length):
    """
    Pads a string of bits with zeros until
    the desired string length, zeros are padded
    to the left of the string.
    :param bit_str: The initial string of bits
    :param target_length: The target length of the bit string
    :return: The result of appending zeros to the bit string
    """
    while len(bit_str) < target_length:
        bit_str = "0" + bit_str
    return bit_str


def pad_zeros_right(bit_str, target_length):
    """
    Pads a string of bits with zeros until
    the desired string length, zeros are padded
    to the right of the string.
    :param bit_str: The initial string of bits
    :param target_length: The target length of the bit string
    :return: The result of appending zeros to the bit string
    """
    while len(bit_str) < target_length:
        bit_str += "0"
    return bit_str


def circular_shift_left(bit_str, num_rounds=1):
    """
    Does circular shift lefts on the supplied
    bit string "num_rounds" times.
    :param bit_str: The initial string of bits
    :param num_rounds: The number of circular shift rounds
    :return: The result of performing the circular shifts
    """
    while num_rounds > 0:
        bit_str = bit_str[1:] + bit_str[0]
        num_rounds -= 1
    return bit_str


def circular_shift_right(bit_str, num_rounds=1):
    """
    Does circular shift rights on the supplied
    bit string "num_rounds" times.
    :param bit_str: The initial string of bits
    :param num_rounds: The number of circular shift rounds
    :return: The result of performing the circular shifts
    """
    while num_rounds > 0:
        bit_str = bit_str[-1] + bit_str[:-1]
        num_rounds -= 1
    return bit_str


def xor(bit_str_x, bit_str_y):
    """
    Performs a logical XOR operation on the supplied
    bit strings.
    :param bit_str_x: The first bit string
    :param bit_str_y: The second bit string
    :return: The result of XORing the two bit strings
    """
    result = ""
    for x_char, y_char in zip(bit_str_x, bit_str_y):
        result += "0" if x_char == y_char else "1"
    return result


def to_bytes(bit_str):
    """
    Converts the supplied bit string to an equivalent
    string of bytes. Assumes string lengths that are
    multiples of 8.
    :param bit_str: The string of bits
    :return: An equivalent string of bytes
    """
    if len(bit_str) % 8 != 0:
        return None
    result = ""
    out_len = int(len(bit_str) / 8)
    for i in range(0, out_len):
        result += chr(int(bit_str[i * 8: (i + 1) * 8], 2))
    return result


def to_binary(byte_str):
    """
    Turns a byte string into a bit string by converting
    every byte to its equivalent binary representation.
    :param byte_str: The string of bytes
    :return: An equivalent string of bits
    """
    result = ""
    for byte in byte_str:
        result += bin(int.from_bytes(byte.encode(), 'big'))[2:].zfill(8)
    return result


def int_to_binary(int):
    """
    Converts an integer to its 8-bit equivalent
    in binary format.
    :param int: The integer to convert
    :return: The binary equivalent
    """
    return bin(int)[2:].zfill(8)
